### DISCUSSION TITLE
Request for Assistance - Achieving Internal Rewriting Without Reflecting URL Changes in the Browser Address Bar

### DISCUSSION THREAD

**Discussion by yashpriyam** [vercel/next.js#60375] [Help]
### Summary

Hi Next.js Community,

I hope this message finds you well. I am currently working on a project where I need to achieve internal URL rewriting without causing the changes to be reflected in the client's browser address bar during client-side navigation. I have explored various approaches, including the use of NextResponse.rewrite, but the URL changes persist in the browser.

Here's a simplified version of the code I've been working with:

```
// Custom server function
export const handleRewriting = async (req, res) => {
  const { pathname, search } = req.url;

  // Check if the request matches a certain condition
  if (/* Some condition */) {
    // Modify the necessary parts for internal rewriting without changing the URL
    const modifiedUrl = /* Construct the modified URL */;
    
    // Need assistance with achieving internal rewriting without changing the browser's address bar
    // ...
  }

  // Continue with default request handling
  // ...
};

```

I have explored options such as using NextResponse.rewrite and fetching the modified content, but the URL changes still appear in the browser address bar. I am reaching out to the community for any insights, guidance, or alternative approaches that could help me achieve internal rewriting without affecting the URL visible to the client.

Your expertise and suggestions would be highly appreciated!

Thank you for your time.

### Example

```
// Custom server function
export const handleRewriting = async (req, res) => {
  const { pathname, search } = req.url;

  // Check if the request matches a certain condition
  if (/* Some condition */) {
    // Modify the necessary parts for internal rewriting without changing the URL
    const modifiedUrl = /* Construct the modified URL */;
    
    // Need assistance with achieving internal rewriting without changing the browser's address bar
    // ...
  }

  // Continue with default request handling
  // ...
};

```

**Comment by yashpriyam** [comment]
@toolmantim @sco @chrislloyd @benschwarz

**Comment by don-esteban** [comment]
Hi @yashpriyam,

please reproduce the error with minimal code.

Perhaps it helps, if I provide a minimal working example:

```
// project-root/middleware.js

import { NextResponse } from 'next/server'
 
export function middleware(request) {
  return NextResponse.rewrite(new URL('/', request.url))
}
 
export const config = {
  matcher: '/about/',
```

This rewrites the `/about/` request to `/` (without changing the URL in the browser)

See:

- https://nextjs.org/docs/app/building-your-application/routing/middleware
- https://nextjs.org/docs/app/api-reference/functions/next-response#rewrite
